### PR TITLE
controllers: allow replica count to update without updating revision SHA

### PR DESCRIPTION
**Pull request by pidren**
This pull request has following fixes:

Fix 1: allow replica count to update without updating revision SHA
- Any statefulset spec to update without updating revision SHA. This allows statefulset to scale up/down without restarting all components in that statefulset (applies to both hbasemaster and regionservers)
- When revision SHA is calculated, current replica count is used instead of some other arbitrary number. This is to avoid the first time this update is deployed triggering another full restart of all components (since arbitrary count would still be different than whatever the count it is currently the first time), but rather, make this change to SHA calculation entirely seamless.

Fix 2: test refactoring
Did some test refactoring mostly to make understanding of the code easier. Tests that were failing because they were immediate assertions rather than an Eventually() test -- sometimes the reconciler simply is slow and makes such tests super flakey - these have been fixed to use Eventually() since we really don't care when reconciler finishes ... reconciling ... as long as it does at some point, within the timeout range.

### CODE REVIEW

File: controllers/hbase_controller.go [hunk 1]
```diff
@@ -628,11 +634,14 @@ const (
...
-	stsName, cmName types.NamespacedName, ss hbasev1.ServerSpec) (*appsv1.StatefulSet, string) {
+	stsName, cmName types.NamespacedName, ss hbasev1.ServerSpec, rc int32) (*appsv1.StatefulSet, string) {
 	spec := (&ss.PodSpec).DeepCopy()
 	spec.Volumes = append(spec.Volumes, configMapVolume(cmName))
 	stsSpec := appsv1.StatefulSetSpec{
-		Replicas:            pointer.Int32Ptr(ss.Count),
+		// We don't want to use replica value as part of revision change (since
+		// scaling a statefulset is not the same as a new revision) so we just
+		// set current count for now just for hash calculation.
+		Replicas:            pointer.Int32Ptr(rc),
```
Comment (timoha) [inline controllers/hbase_controller.go:644]: why not just leave it empty?

File: controllers/hbase_controller.go [hunk 2]
```diff
@@ -564,9 +563,10 @@ func (r *HBaseReconciler) ensureStatefulSetPods(ctx context.Context, sts *appsv1
 func (r *HBaseReconciler) ensureStatefulSet(hb *hbasev1.HBase,
 	stsName, cmName types.NamespacedName, ss hbasev1.ServerSpec) (*appsv1.StatefulSet, bool, error) {
 	actual := &appsv1.StatefulSet{}
-	expected, expectedRevision := r.statefulSet(hb, stsName, cmName, ss)
 	if err := r.Get(context.TODO(), stsName, actual); err != nil {
 		if errors.IsNotFound(err) {
+			expected, _ := r.statefulSet(hb, stsName, cmName, ss, 0)
```
Comment (timoha) [inline controllers/hbase_controller.go:568]: what's the reason for this change?

File: controllers/hbase_controller.go [hunk 3]
```diff
@@ -578,8 +578,14 @@ func (r *HBaseReconciler) ensureStatefulSet(hb *hbasev1.HBase,
 		}
 		return nil, false, err
 	}
+	expected, expectedRevision := r.statefulSet(hb, stsName, cmName, ss, *actual.Spec.Replicas)
+
+	r.Log.Info("updating revision", "?", expectedRevision != actual.Annotations[HBaseControllerRevisionKey])
```
Comment (timoha) [inline controllers/hbase_controller.go:583]: what is `?` as second arg to Info?

File: controllers/hbase_controller.go [hunk 4]
```diff
@@ -491,9 +489,10 @@ func (r *HBaseReconciler) ensureStatefulSetPods(ctx context.Context, sts *appsv1
 
 	r.Log.Info("matched pods", "statefulset", sts.Name, "pods", len(podList.Items))
 
-	// sort pods by name to have things restarted predictably
+	// sort pods in reverse ordinal order to restart
+	//https://kubernetes.io/docs/tutorials/stateful-application/basic-stateful-set/#rolling-update
 	sort.Slice(podList.Items, func(i, j int) bool {
-		return podList.Items[i].Name < podList.Items[j].Name
+		return podList.Items[i].Name > podList.Items[j].Name
```
Comment (timoha) [inline controllers/hbase_controller.go:495]: since it's comparing based on name, it still won't restart similar to how statefulsets are restarted. Because sorting would be like so for anything past single digit:

regionserver-9
...
regionserver-2
regionserver-11
regionserver-10
regionserver-1

File: go.mod [hunk 5]
```diff
@@ -11,6 +11,7 @@ require (
 	github.com/konsorten/go-windows-terminal-sequences v1.0.3 // indirect
 	github.com/onsi/ginkgo v1.12.1
 	github.com/onsi/gomega v1.10.1
+	github.com/prometheus/common v0.4.1
```
Comment (timoha) [inline go.mod:14]: where's this dependency used?

File: controllers/hbase_controller.go (hunk 1 above)
Comment (pidren) [reply controllers/hbase_controller.go:644]: Because leaving it empty still forces a SHA changes the first time this code is deployed, causing all components to restart, even if it's just once and never again. From commit msg:

"When revision SHA is calculated, current replica count is used instead of some other arbitrary number. This is to avoid the first time this update is deployed triggering another full restart of all components (since arbitrary count would still be different than whatever the count it is currently the first time), but rather, make this change to SHA calculation entirely seamless."

File: controllers/hbase_controller.go (hunk 3 above)
Comment (pidren) [reply controllers/hbase_controller.go:583]: Honestly, just a bogus filler. This reconciler log requires 3 arguments, and in some of the code we use "name" as a string arg for the 2nd. I can revert to using "name", it just made reading the logs easier to have "?" because it reads like:

2021-05-03T09:42:12.067-0700 INFO controllers.HBase updating revision {"?": false}

File: controllers/hbase_controller.go (hunk 4 above)
Comment (pidren) [reply controllers/hbase_controller.go:495]: Thanks, good point. Let me remove this commit from this pull request, since it's unrelated to the SHA fix and add test cases for ensureStatefulSetPods()

File: controllers/hbase_controller.go (hunk 3 above)
Comment (timoha) [reply controllers/hbase_controller.go:583]: can't you call it something that it actually is? like in this case `sameRevision: false`

File: controllers/hbase_controller.go (hunk 1 above)
Comment (timoha) [reply controllers/hbase_controller.go:644]: sounds good, sorry that didn't check the review commit

File: controllers/hbase_controller.go (hunk 2 above)
Comment (pidren) [reply controllers/hbase_controller.go:568]: This is handling the case where statefulset is created for the first time, so for that the existing count would be 0 (best representative of "does not exist")

File: controllers/hbase_controller.go (hunk 3 above)
Comment (pidren) [reply controllers/hbase_controller.go:583]: ACK

File: go.mod (hunk 5 above)
Comment (pidren) [reply go.mod:14]: Not a required dependency, removed

**Review by timoha — APPROVED** [review]
